index.js

```diff
@@ -2995,66 +3135,241 @@ function Env(handle) {
     };
 })();
 
-let _artMethodSpec = null;
-function getArtMethodSpec(vm) {
-    if (_artMethodSpec !== null)
-        return _artMethodSpec;
-
-    vm.perform(() => {
-        const env = vm.getEnv();
-        const process = env.findClass("android/os/Process");
-        const setArgV0 = env.getStaticMethodId(process, "setArgV0", "(Ljava/lang/String;)V");
+function resolveStructSpec(vm) {
+    if (this._current !== undefined)
+        return this._current;
 
-        const runtimeModule = Process.getModuleByName("libandroid_runtime.so");
-        const runtimeStart = runtimeModule.base;
-        const runtimeEnd = runtimeStart.add(runtimeModule.size);
+    let specByPointerSize = this[runtime.androidVersion.substr(0, 4)];
+    if (specByPointerSize === undefined) {
+        if (this._default) {
+            let spec = this._default(vm);
+            if (k) {
+                this._current = spec;
+                return spec;
+            }
+        }
+        specByPointerSize = this.git;
+    }
 
-        const expectedAccessFlags = kAccPublic | kAccStatic | kAccFinal | kAccNative;
+    this._current = specByPointerSize && specByPointerSize[Process.pointerSize];
+    return this._current;
+}
 
-        let jniCodeOffset = -1;
-        let accessFlagsOffset = -1;
-        let remaining = 2;
-        for (let offset = 0; offset !== 64 && remaining !== 0; offset += 4) {
-            const field = setArgV0.add(offset);
+const getArtMethodSpec = resolveStructSpec.bind({
```

Comment:
@iamahuman FWIW I have now split index.js into multiple modules, and implemented calculation of all the fields:
https://github.com/frida/frida-java/blob/6f93ee08924231dae50565511d1ad1532055eedd/lib/android.js

This test is now passing with four test-devices connected -- three of which are 32-bit ARM emulators running 5.0, 5.1, and 6.0, and the last of which is a 64-bit Samsung Galaxy S6 Edge device:
https://github.com/frida/frida-java/blob/6f93ee08924231dae50565511d1ad1532055eedd/test/android.js#L22-L111
(Yeah the test is a bit verbose, but at least it's very explicit.)

Some testing still remains, but should hopefully be able to release this soon.